### PR TITLE
RCHAIN-4077: Key-value store with manager

### PR DESCRIPTION
**Pull request by tgrospic**
## Overview
This PR introduces key-value store abstraction with the manager for easier and general use than existing `Store`.
Manager gives more flexibility how implementation deals with database connections and environments. We can manage all store instances in one place (node project) and relief other places (like RSpace) from knowing about implementation details. After migration, LMDB can be removed as dependency from other projects (shared, block-storage and rspace) and can be moved to node project (and main test project).
This will also enable much easier integration and testing with other key-value databases like RocksDB or Badger.

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-4077

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: node/src/main/scala/coop/rchain/node/storage/RNodeKeyValueStoreManager.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  // Database name to store instance name mapping (subfolder for LMDB store)
+  // - keys with the same instance will be in one LMDB file (environment)
+  val instanceDBmapping: Map[String, String] = Map[String, String](
+    // RSpace
+    ("history", "v2/history"),
+    ("cold", "v2/cold"),
+    ("roots", "v2/roots"),
+    // Block store
+    ("block-metadata", "dagstorage")
```
Comment (tgrospic) [inline node/src/main/scala/coop/rchain/node/storage/RNodeKeyValueStoreManager.scala]: `RNodeKeyValueStoreManager` should accept Config and so that this mapping is exactly created from config parameters.

**Comment by nzpr** [comment]
@tgrospic will you add `remove` method to KeyValueStore trait in this PR?

**Comment by tgrospic** [comment]
@nzpr, @zsluedem PR is ready for review. I've added `delete` to KeyValueStore and tests.
Base branch is `dev` so we can use it for trace persistent cache.

File: casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  // Create LMDB environment and update the state
+  private def createEnv: F[Unit] =
+    for {
+      _     <- Sync[F].delay(Files.createDirectories(dirPath))
+      flags = Seq(EnvFlags.MDB_NOTLS)
+      // Create environment
+      env <- Sync[F].delay(
+              Env
+                .create()
+                .setMapSize(10L * 1024L * 1024L * 1024L)
```
Comment (zsluedem) [inline casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala:116]: Makes the param configurable is prefered for me. But we can do that in another pr. Maybe create a ticket for that

File: casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      // Find DB ref / first time create deferred object.
+      newDbDefer <- Deferred[F, DbEnv[F]]
+      action <- varState.modify { st =>
+                 val maybeDbRef = st.dbs.get(dbName)
+                 maybeDbRef.fold {
+                   val newDbs = st.dbs.updated(dbName, newDbDefer)
+                   val newSt  = st.copy(dbs = newDbs)
+                   (newSt, (true, newDbDefer))
+                 }(defer => (st, (false, defer)))
+               }
```
Comment (zsluedem) [inline casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala:76]: So each defer in each db can be a guarantee that the db can be used only if it is created, right?
It looks like the `true` `false` are dups? Not sure if i am right.

**Review by zsluedem — COMMENTED** [review]
Very good!! Leave some comments!

File: casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala (hunk 2 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala:116]: @zsluedem the idea is for RNodeKeyValueManager to accepts configuration for all stores and instantiate LmdbStoreManager parameters like map size or database count. We can do this as part of PR where it will be used first.

File: casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala (hunk 3 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala:76]: @zsluedem yes, each caller will be blocked with Deferred until it's completed which is done when DB is created. This is the first step to implement reload of the environment where we need to wait operations on existing environment to finish.

Returning a flag if new environment should be created is a way to capture transition that must happen only once. Only the first request should trigger creation of a new database. But this "effect" cannot be executed inside Ref `modify` function because this function should only depend on the internal state and can be executed multiple times if the state is changed between reading and writing. So the flag `true` will be only signaled once per database for "effectful" code to execute.

**Review by nzpr — APPROVED** [review]
Great work! Beautiful!

**Review by zsluedem — APPROVED** [review]
